alacritty_terminal/src/grid/mod.rs

```diff
@@ -263,46 +287,57 @@ impl<T: GridCell + Default + PartialEq + Clone> Grid<T> {
-            if self.display_offset != 0 {
-                self.display_offset = min(self.display_offset + *positions, self.max_scroll_limit);
+        // When rotating the entire region with fixed lines at the top, just reset everything.
+        if positions >= region.end - region.start && region.start != Line(0) {
+            for i in region.start.0..region.end.0 {
+                let index = screen_lines - i - 1;
+                self.raw[index].reset(&self.cursor.template);
             }
 
-            self.increase_scroll_limit(*positions);
+            return;
+        }
 
-            // Rotate the entire line buffer. If there's a scrolling region
-            // active, the bottom lines are restored in the next step.
-            self.raw.rotate(-(*positions as isize));
+        // Update display offset when not pinned to active area.
+        if self.display_offset != 0 {
+            self.display_offset = min(self.display_offset + *positions, self.max_scroll_limit);
+        }
 
-            // This next loop swaps "fixed" lines outside of a scroll region
-            // back into place after the rotation. The work is done in buffer-
-            // space rather than terminal-space to avoid redundant
-            // transformations.
-            let fixed_lines = num_lines - *region.end;
+        // Create scrollback for the new lines.
+        self.increase_scroll_limit(*positions);
 
-            for i in 0..fixed_lines {
-                self.raw.swap(i, i + *positions);
-            }
+        // Swap the lines fixed at the top to their target positions after rotation.
+        //
+        // Since we've made sure that the rotation will never rotate away the entire region, we
+        // know that the position of the fixed lines before the rotation must already be
+        // visible.
+        //
+        // We need to start from the bottom, to make sure the fixed lines aren't swapped with each
+        // other.
+        for i in (0..region.start.0).rev() {
+            let index = screen_lines - i - 1;
+            self.raw.swap(index, index - positions.0);
+        }
 
-            // Finally, reset recycled lines.
-            //
-            // Recycled lines are just above the end of the scrolling region.
-            for i in 0..*positions {
-                self.raw[i + fixed_lines].reset(&self.cursor.template);
-            }
-        } else {
-            // Subregion rotation.
-            for line in IndexRange(region.start..(region.end - positions)) {
-                self.raw.swap_lines(line, line + positions);
-            }
+        // TODO: Technically the rotation here will rotate dirty fixed lines at the top of the
+        // viewport into history. Should these be cleared? This is only a problem when rotating
+        // more than a single line.
+        //  -> I say it doesn't matter, since you shouldn't use a scrolling region like that
+        //  outside of the alt screen anyways.
```

Comment:
@kchibisov This comment talks about the same thing we've discussed on IRC. Generally I'd say that the performance is significantly better for the purpose of this PR, while not touching the performance in other areas much.

Do you think there's some other approach that should be evaluated? If so, what would you suggest? I've tried to create a single thing that fixes everything, but if we wanted to implement multiple different variations we'd probably need at least three. One optimized for primary screen scrolling, one for alt screen with the fixed lines and then one fallback that can handle everything but is slow.